docs/proposals/cascading-deletion.md

```diff
@@ -0,0 +1,230 @@
+<!-- BEGIN MUNGE: UNVERSIONED_WARNING -->
+
+<!-- BEGIN STRIP_FOR_RELEASE -->
+
+<img src="http://kubernetes.io/img/warning.png" alt="WARNING"
+     width="25" height="25">
+<img src="http://kubernetes.io/img/warning.png" alt="WARNING"
+     width="25" height="25">
+<img src="http://kubernetes.io/img/warning.png" alt="WARNING"
+     width="25" height="25">
+<img src="http://kubernetes.io/img/warning.png" alt="WARNING"
+     width="25" height="25">
+<img src="http://kubernetes.io/img/warning.png" alt="WARNING"
+     width="25" height="25">
+
+<h2>PLEASE NOTE: This document applies to the HEAD of the source tree</h2>
+
+If you are using a released version of Kubernetes, you should
+refer to the docs that go with that version.
+
+Documentation for other releases can be found at
+[releases.k8s.io](http://releases.k8s.io).
+</strong>
+--
+
+<!-- END STRIP_FOR_RELEASE -->
+
+<!-- END MUNGE: UNVERSIONED_WARNING -->
+
+# Overview
+
+Currently most cascading deletion logic is implemented at client-side. For example, when deleting a replica set, kubectl uses a reaper to delete the created pods and then delete the replica set. We plan to move the cascading deletion to the server to simplify the client-side logic. In this proposal, we present two alternative designs. [Design #1](#design-1-tombstone--gc) uses the tombstone mechanism, and [Desig #2](#design-2-finalizer-framework--gc) uses the finalizer mechanism. Some
+common design choices faced by both designs are described in [common designs](#common-designs).
+
+# Goals
+
+* Supporting cascading deletion at the server-side.
+* Centralizing the cascading deletion logic, rather than spreading in controllers.
+* Allowing optionally orphan the dependent objects.
+
+# Non goals
+
+* Releasing the name of an object immediately, so it can be reused ASAP.
+
+# Design 1. Tombstone + GC
+
+## API Changes
+
+```
+type DeleteOptions struct {
+	…
+	OrphanChildren bool
+}
+```
+
+**DeleteOptions.OrphanChildren**: allows a user to express whether the child objects should be orphaned.
+
+```
+type ObjectMeta struct {
+	...	
+	ParentReferences []ObjectReference
+}
+```
+
+**ObjectMeta.ParentReferences**: links the resource to the parent resources. For example, a replica set `R` created by a deployment `D` should have an entry in ObjectMeta.ParentReferences pointing to `D`. The link should be set when the child object is created. It can be updated after the creation.
```

Comment:
I'm not sure I follow, can you give an example?

On Tue, Apr 12, 2016 at 8:30 PM, Derek Carr <notifications@github.com>
wrote:

> In docs/proposals/cascading-deletion.md
> <https://github.com/kubernetes/kubernetes/pull/23656#discussion_r59488717>
> :
>
> > +type DeleteOptions struct {
> > +	…
> > +	OrphanChildren bool
> > +}
> > +```
> > +
> > +**DeleteOptions.OrphanChildren**: allows a user to express whether the child objects should be orphaned.
> > +
> > +```
> > +type ObjectMeta struct {
> > +	...	
> > +	ParentReferences []ObjectReference
> > +}
> > +```
> > +
> > +**ObjectMeta.ParentReferences**: links the resource to the parent resources. For example, a replica set `R` created by a deployment `D` should have an entry in ObjectMeta.ParentReferences pointing to `D`. The link should be set when the child object is created. It can be updated after the creation.
>
> The current Namespace finalizer is similar to the dependent object, but in
> this case the dependency is a cluster provider of some kind (Kubernetes,
> OpenShift). If we go this route for dependencies as a object reference, we
> need to let non-namespaced resources depend on something akin to a API
> provider that must be registered with the cluster. I also could see the
> same value in Node(s)
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly or view it on GitHub
> <https://github.com/kubernetes/kubernetes/pull/23656/files/cc943fae771b704370a7ddca1d6267b6786fd807#r59488717>
>